HeterogeneousCore/AlpakaMath/interface/Vector.h

```diff
@@ -0,0 +1,301 @@
+    Vector<T, maxSize> &operator=(Vector<T, maxSize> &&) = default;
+
+    constexpr T const *begin() const { return m_data; }
+    constexpr T const *end() const { return m_data + m_size; }
+    constexpr T *begin() { return m_data; }
+    constexpr T *end() { return m_data + m_size; }
+    constexpr int size() const { return m_size; }
+    constexpr T &operator[](int i) { return m_data[i]; }
+    constexpr const T &operator[](int i) const { return m_data[i]; }
+    constexpr void reset() { m_size = 0; }
+    static constexpr int capacity() { return maxSize; }
+    constexpr T const *data() const { return m_data; }
+    constexpr void resize(int size) { m_size = size; }
+    constexpr bool empty() const { return 0 == m_size; }
+    constexpr bool full() const { return maxSize == m_size; }
+
+    constexpr void zero() {
+      CMS_UNROLL_LOOP
+      for (int i = 0; i < maxSize; i++)
+        m_data[i] = zero_<T>();
+    }
+
+    constexpr T norm2() const {
+      T res{0};
+      CMS_UNROLL_LOOP
+      for (int i = 0; i < maxSize; i++) {
+        res += m_data[i] * m_data[i];
+      }
+      return res;
+    }
+
+    template <unsigned int n>
+    constexpr T partial_norm2() const {
+      constexpr int reduced_size = n > maxSize ? maxSize : n;
+
+      T res{0};
+      CMS_UNROLL_LOOP
+      for (int i = 0; i < reduced_size; i++) {
+        res += m_data[i] * m_data[i];
+      }
+
+      return res;
+    }
+
+    constexpr Vector<T, maxSize> &operator*=(const T &scale) {
+      CMS_UNROLL_LOOP
+      for (int i = 0; i < maxSize; ++i) {
+        m_data[i] *= scale;
+      }
+      return *this;
+    }
+
+    constexpr Vector<T, maxSize> &operator+=(const Vector<T, maxSize> &rhs) {
+      CMS_UNROLL_LOOP
+      for (int i = 0; i < maxSize; ++i) {
+        m_data[i] += rhs[i];
+      }
+      return *this;
+    }
+
+    template <typename TAcc>
+    ALPAKA_FN_ACC T norm(const TAcc &acc) const {
+      const T nrm2 = norm2();
+      return alpaka::math::sqrt(acc, nrm2);
+    }
+
+    template <typename TAcc, unsigned int n>
+    ALPAKA_FN_ACC T partial_norm(const TAcc &acc) const {
+      const T partial_nrm2 = partial_norm2<n>();
+      return alpaka::math::sqrt(acc, partial_nrm2);
+    }
+
+    template <typename TAcc>
+    ALPAKA_FN_ACC void normalize(const TAcc &acc) {
+      const T nrm = norm(acc);
+
+      if (nrm == static_cast<T>(0))
+        return;
+
+      CMS_UNROLL_LOOP
+      for (int i = 0; i < maxSize; i++) {
+        m_data[i] /= nrm;
+      }
+    }
+
+  private:
+    T m_data[maxSize];
+
+    int m_size;
+  };
+
+  template <typename T>
+  struct is_Vector : std::false_type {};
+
+  template <typename T, int N>
+  struct is_Vector<cms::alpakatools::math::Vector<T, N>> : std::true_type {};
+
+  template <typename T>
+  inline constexpr bool is_Vector_v = is_Vector<T>::value;
+
+  template <typename VectorN, std::enable_if_t<is_Vector_v<VectorN>, int> = 0>
+  inline constexpr VectorN zero() {
```

Comment:
How about we make the default constructor to initialize the data with zeros?

Actually, with `m_data{}` doesn't the default constructor already initialize the data with zeros?